relay-server/src/services/spooler/envelope_stack/sqlite.rs

```diff
@@ -85,46 +90,78 @@ impl SQLiteEnvelopeStack {
+        if self.batches_buffer_size > 0 {
+            return Ok(());
+        }
+
+        let Some(spooling_handle) = mem::take(&mut self.spooling_handle) else {
+            return Ok(());
+        };
+
+        spooling_handle
+            .await
+            .map_err(|_| SQLiteEnvelopeStackError::AsyncSpoolingError)
+    }
+
     /// Spools to disk up to `disk_batch_size` envelopes from the `buffer`.
     ///
     /// In case there is a failure while writing envelopes, all the envelopes that were enqueued
     /// to be written to disk are lost. The explanation for this behavior can be found in the body
     /// of the method.
-    async fn spool_to_disk(&mut self) -> Result<(), SQLiteEnvelopeStackError> {
+    fn background_spool_to_disk(&mut self) {
         let Some(envelopes) = self.batches_buffer.pop_front() else {
-            return Ok(());
+            return;
         };
+        // We optimistically assume that such elements were removed from the buffer, so we update
+        // the new size.
         self.batches_buffer_size -= envelopes.len();
 
-        let insert_envelopes = envelopes.iter().map(|e| InsertEnvelope {
-            received_at: received_at(e),
-            own_key: self.own_key,
-            sampling_key: self.sampling_key,
-            encoded_envelope: e.to_vec().unwrap(),
+        let db = self.db.clone();
+        let own_key = self.own_key;
+        let sampling_key = self.sampling_key;
+        let spooling_handle = tokio::spawn(async move {
+            let insert_envelopes = envelopes.iter().map(|e| InsertEnvelope {
+                received_at: received_at(e),
+                own_key,
+                sampling_key,
+                encoded_envelope: e.to_vec().unwrap(),
+            });
+
+            // When spawning this task, we are acknowledging that the elements that we popped from
+            // the buffer are lost in case of error during insertion.
+            //
+            // We are doing this on purposes, since if we were to have a database corruption during
+            // runtime, and we were to put the values back into the buffer we will end up with an
+            // infinite cycle.
+            if let Err(err) = build_insert_many_envelopes(insert_envelopes)
+                .build()
+                .execute(&db)
+                .await
+            {
+                relay_log::error!(
+                    error = &err as &dyn Error,
+                    own_key = own_key.to_string(),
+                    sampling_key = sampling_key.to_string(),
+                    number_of_envelopes = envelopes.len(),
+                    "failed to spool envelopes to disk",
+                );
+            }
         });
+        // TODO: we might need to revisit this code to make sure that the handle has finished before
+        //  potentially shutting down the runtime/task.
+        self.spooling_handle = Some(spooling_handle);
```

Comment:
Can it happen that the spool handle is already `Some` here?

```suggestion
        debug_assert!(spooling_handle.is_none());
        self.spooling_handle = Some(spooling_handle);
```